[SPARK-47318][CORE] Adds HKDF round to AuthEngine key derivation to follow standard KEX practices

sweisdb opened the pull request:
### What changes were proposed in this pull request?

This change adds an additional pass through a key derivation function (KDF) to the key exchange protocol in `AuthEngine`. Currently, it uses the shared secret from a bespoke key negotiation protocol directly. This is an encoded X coordinate on the X25519 curve. It is atypical and not recommended to use that coordinate directly as a key, but rather to pass it to an KDF.

Note, Spark now supports TLS for RPC calls. It is preferable to use that rather than the bespoke AES RPC encryption implemented by `AuthEngine` and `TransportCipher`.

### Why are the changes needed?

This follows best practices of key negotiation protocols. The encoded X coordinate is not guaranteed to be uniformly distributed over the 32-byte key space. Rather, we pass it through a HKDF function to map it uniformly to a 16-byte key space.

### Does this PR introduce _any_ user-facing change?

No.

### How was this patch tested?

Exiting tests under:
`build/sbt "network-common/test:testOnly"`

Specifically:
`build/sbt "network-common/test:testOnly org.apache.spark.network.crypto.AuthEngineSuite"`
`build/sbt "network-common/test:testOnly org.apache.spark.network.crypto.AuthIntegrationSuite"`

### Was this patch authored or co-authored using generative AI tooling?

No.

sweisdb committed:
[SPARK-47318][Security] Adds HKDF round to AuthEngine key derivation

mridulm commented:
+CC @vanzin, @srowen 

Can you also include the documentation change in the PR to call out the incompatibility introduced due to this ? Thanks

dongjoon-hyun reviewed (commented):
Could you add a reference link for this statement in the PR description, @sweisdb ?
> This follows best practices of key negotiation protocols.

sweisdb commented:
@dongjoon-hyun Will do. I would clarify that the best practice here would be to use TLS and not use a bespoke key exchange protocol. 

But if we are going to use a bespoke key exchange protocol, we need to run the output of the key exchange through a KDF.

sweisdb committed:
[SPARK-47318] Updating documentation for AuthEngine KEX change

sweisdb commented:
I've updated the overall Security doc to indicate that TLS is preferred to using the legacy RPC encryption. I also updated the AuthEngine README to talk about incompatibility.

I retroactively made up v1.0 and v1.1 of the key exchange protocol to talk about this. It would probably be more understandable to talk about Spark versions. Should we talk about 4.0 in the context of this change and amend the docs if and when it's backported?

mridulm commented on common/network-common/src/main/java/org/apache/spark/network/crypto/README.md:
```diff
@@ -1,6 +1,20 @@
+
+See
+the [Spark security documentation](https://github.com/apache/spark/blob/master/docs/security.md#ssl-encryption) for
+more information on how to configure TLS.
```
I am not sure if we should announce deprecation given TLS is going to come out only in 4.0 ... we can recommend users to move to TLS though.

Would like feedback from others on this though ... @dongjoon-hyun, @srowen, @JoshRosen

sweisdb replied:
Given that SASL support still exists for backward compatibility, I am skeptical that this will actually be removed. I'll delete that comment.

mridulm commented on common/network-common/src/main/java/org/apache/spark/network/crypto/README.md:
```diff
@@ -99,3 +114,13 @@ sessions. It would, however, allow impersonation of future sessions.
+The original version of this protocol, retroactively called v1.0, did not apply an HKDF to `sharedSecret` and was
+directly using the encoded X coordinate as key material. This is atypical and standard practice is to pass that shared
+coordinate through an HKDF. The current version, v1.1, adds this additional HKDF to
+derive `derivedKey`.
```
Essentially, this is backwardly incompatible with 3.x  on the wire - making adoption of 4.x in secure env all-or-nothing
Given this, I am leaning towards flag guarding the change.

dongjoon-hyun replied:
+1 for @mridulm 's recommendation.

mridulm commented on common/network-common/src/main/java/org/apache/spark/network/crypto/AuthEngine.java:
```diff
@@ -213,7 +220,7 @@ private TransportCipher generateTransportCipher(
         OUTPUT_IV_INFO,  // This is the HKDF info field used to differentiate IV values
         AES_GCM_KEY_SIZE_BYTES);
-    SecretKeySpec sessionKey = new SecretKeySpec(sharedSecret, "AES");
+    SecretKeySpec sessionKey = new SecretKeySpec(derivedKey, "AES");
```
Let us flag guard this change - given we want to backport to 3.x patch releases as well as giving users migration opportunities. This flag can be flipped to default in a subsequent minor version.

sweisdb replied:
I added a flag to let someone override this and revert back to the original protocol.

mridulm reviewed (commented):
For context to reviewers, the plan is to backport this to 3.x as well given security implications - that is, we will will be making an incompatible wire protocol change.

Please review from that context as well - thanks !

dongjoon-hyun commented on common/network-common/src/main/java/org/apache/spark/network/crypto/README.md:
```diff
@@ -99,3 +114,13 @@ sessions. It would, however, allow impersonation of future sessions.
+derive `derivedKey`.
+
+Consequently, older Spark versions using v1.0 of this protocol will not negotiate the same key as
+Spark versions using v1.1 and will be **unable to send encrypted RPCs** across incompatible versions.
```
I assume that Apache Spark community follows `Semantic Versioning` by default. If this is incompatible, shall we use `v2.0` instead of `v1.1`?

dongjoon-hyun commented on common/network-common/src/main/java/org/apache/spark/network/crypto/README.md:
```diff
@@ -1,6 +1,20 @@
-Forward Secure Auth Protocol
+Forward Secure Auth Protocol v1.1
```
`v2.0`? (https://github.com/apache/spark/pull/45425/files/4204dd1f03cb86f4f6e082d2c4b8bd5f402ad12f..7ace803c40ab50c03e4af1c3733499668fb340d5#r1525394252)

sweisdb replied:
Will fix.

dongjoon-hyun commented on docs/security.md:
```diff
@@ -149,24 +149,32 @@ secret file agrees with the executors' secret file.
+The RPC implementation uses Netty under the hood (while the UI uses Jetty), which supports a
+different set of options.
+
+Unlike the other SSL settings for the UI, the RPC SSL is *not* automatically enabled if
```
Since this layer is completely irrelevant to UI, we don't need to say like this. Shall we omit this paragraph?
> Unlike the other SSL settings for the UI

dongjoon-hyun commented on common/network-common/src/main/java/org/apache/spark/network/crypto/README.md:
```diff
@@ -1,6 +1,20 @@
+------------------
+This is a bespoke key exchange protocol that was implemented before Spark supported TLS (aka SSL) for RPC
+calls. It is recommended that Spark users upgrade to using TLS for RPC calls between Spark processes. This protocol
+will be deprecated and removed in the long-term.
```
To @sweisdb , may I ask when you think we can delete this? If you think `Spark 5`, please remove this claim. It's too far. In Apache Spark community, deprecation doesn't mean the removal.
> and removed in the long-term.

FYI, here is the guideline.
- https://spark.apache.org/versioning-policy.html
https://spark.apache.org/versioning-policy.html

sweisdb replied:
After seeing that we're still supporting legacy SASL, I agree that this will likely not be removed any time in the forseeable future.

sweisdb replied:
Fixing this.

dongjoon-hyun reviewed (commented):
BTW, from my side, we don't have a mixed Spark-version RPC use case. So, it's okay to introduce this if we have the fallback configuration to disable this completely (as recommended by Mridul).

sweisdb committed:
[SPARK-47318] Updating Crypto docs

dongjoon-hyun commented:
Got it, @mridulm .
> For context to reviewers, the plan is to backport this to 3.x as well given security implications - that is, we will will be making an incompatible wire protocol change.

If this is for Apache Spark 3.5.2 and 3.4.3, cc @cloud-fan , @HyukjinKwon , too

sweisdb committed:
[SPARK-47318] Adding a flag to support old AuthEngine protocol

sweisdb commented:
I had a bad push that pulled in a lot of other unrelated commits. Will fix.

mridulm commented on common/network-common/src/main/java/org/apache/spark/network/util/TransportConf.java:
```diff
@@ -213,6 +213,11 @@ public boolean encryptionEnabled() {
+  /**
+   * Version number to be used by the AuthEngine key agreement protocol. Value values are 1 or 2.
+   */
+  public int authEngineVersion() { return conf.getInt("spark.network.crypto.authEngineVersion", 2); }
```
Make the default 1, to preserve the existing behavior, and allowing backport to patch versions.
We can document to set this explicitly to to `2` to mitigate this issue.

In a later minor version, we can flip this to `2` by default.

dongjoon-hyun replied:
+1 for the above comment, too.

mridulm commented on common/network-common/src/main/java/org/apache/spark/network/crypto/AuthEngine.java:
```diff
@@ -224,7 +236,7 @@ private TransportCipher generateTransportCipher(
   private byte[] getTranscript(AuthMessage... encryptedPublicKeys) {
     ByteBuf transcript = Unpooled.buffer(
-        Arrays.stream(encryptedPublicKeys).mapToInt(k -> k.encodedLength()).sum());
+        Arrays.stream(encryptedPublicKeys).mapToInt(AuthMessage::encodedLength).sum());
```
Let us minimize optional changes in this PR (this by itself is fine - but in general, we should reduce the scope of change), given we want to backport to older versions and want to minimize risk.

dongjoon-hyun replied:
+1 for @mridulm 's comment. This should be reverted.

sweisdb replied:
Reverted. Good call.

mridulm commented on common/network-common/src/main/java/org/apache/spark/network/crypto/README.md:
```diff
@@ -99,3 +103,13 @@ sessions. It would, however, allow impersonation of future sessions.
+Security Changes & Compatibility
+-------------
+
+The original version of this protocol, retroactively called v1.0, did not apply an HKDF to `sharedSecret` and was
```
nit:

```suggestion
The original version of this protocol, retroactively called v1.0, did not apply an HKDF to `sharedSecret` (to generate the `derivedKey`) and was
```

mridulm commented on docs/security.md:
```diff
@@ -149,24 +149,32 @@ secret file agrees with the executors' secret file.
-the AES based encryption is simpler to configure and may be preferred if the only requirement
-is that data be encrypted in transit.
+The legacy method is an AES-based encryption mechanism relying on a shared secret. This requires
+RPC authentication to also be enabled. This method uses a bespoke protocol and it is recommended
+to use SSL instead.
```
```suggestion
RPC authentication to also be enabled. This method uses a bespoke protocol with an older v1 and an updated v2 version. <add details about v1 vs v2>
```

Also, update the config section below on how to configure `spark.network.crypto.authEngineVersion`, and a ref to that here: along with incompatibility between `1` and `2` (and all versions prior being `1` retroactively - without a way to configure it before `3.5.2` and `3.4.3`)

sweisdb replied:
Updated in the section about the Legacy protocol.

mridulm reviewed (commented):
+CC @dongjoon-hyun as well.

sweisdb committed:
[SPARK-47318] Sets AuthEngine default version to 1 and updates docs

dongjoon-hyun commented on common/network-common/src/main/java/org/apache/spark/network/util/TransportConf.java:
```diff
@@ -213,6 +213,11 @@ public boolean encryptionEnabled() {
   }
 
+  /**
+   * Version number to be used by the AuthEngine key agreement protocol. Value values are 1 or 2.
```
`Value values`?

dongjoon-hyun replied:
Also, shall we mentioned that the default is 1 intentionally?

sweisdb replied:
Good catch. Done.

dongjoon-hyun commented on common/network-common/src/main/java/org/apache/spark/network/crypto/README.md:
```diff
@@ -99,3 +103,13 @@ sessions. It would, however, allow impersonation of future sessions.
+
+The original version of this protocol, retroactively called v1.0, did not apply an HKDF to `sharedSecret` to derive
+a key (i.e. `derivedKey`) and was directly using the encoded X coordinate as key material. This is atypical and
+standard practice is to pass that shared coordinate through an HKDF. The current version, v2.0, adds this additional
```
`The current version, v2.0,` could mislead the users. Let's say `The latest version` instead.

dongjoon-hyun commented on common/network-common/src/main/java/org/apache/spark/network/crypto/README.md:
```diff
@@ -99,3 +103,13 @@ sessions. It would, however, allow impersonation of future sessions.
+HKDF to derive `derivedKey`.
+
+Consequently, older Spark versions using v1.0 of this protocol will not negotiate the same key as
+Spark versions using v2.0 and will be **unable to send encrypted RPCs** across incompatible versions.
```
Shall we add a sentence to describe Apache Spark sticks to `v1.0` by default due to the above reason?

dongjoon-hyun commented on common/network-common/src/test/java/org/apache/spark/network/crypto/AuthEngineSuite.java:
```diff
@@ -48,15 +50,19 @@ public class AuthEngineSuite {
   public static void setUp() {
-    conf = new TransportConf("rpc", MapConfigProvider.EMPTY);
+    ConfigProvider v2Provider = new MapConfigProvider(Collections.singletonMap(
+            "spark.network.crypto.authEngineVersion", "2"));
```
I'm wondering if this means that we don't test `v1` from now.

We had better keep the test coverage for v1 in addition to v2.

sweisdb replied:
We test both v1 and v2. See below where I specifically give it a config with "1".

dongjoon-hyun commented on docs/security.md:
```diff
@@ -149,30 +149,44 @@ secret file agrees with the executors' secret file.
+(KDF) to the key exchange protocol's output, while version 2 applies a KDF to ensure that the derived session
+key is uniformly distributed. Version 1 is default for backward compatibility. It is **recommended to use version 2**
+for better security properties. The version can be configured by setting `spark.network.crypto.authEngineVersion` to
+1 or 2 respectively.
```
Thank you for the above description.

dongjoon-hyun commented on docs/security.md:
```diff
@@ -188,6 +202,11 @@ The following table describes the different options available for configuring th
+<tr>
+  <td><code>spark.network.crypto.authEngineVersion</code></td>
+  <td>1</td>
+  <td>Version of AES-based RPC encryption to use. Valid versions are 1 or 2. Version 2 is recommended. Added in 4.0.0</td>
```
`Added in 4.0.0` looks like a typo. This will break this HTML table. In other words, we need to have two lines of `<td></td>`. One for description and the other for version. Please remove ` Added in 4.0.0` from here and create a new line, `<td>4.0.0</td>`, in order to follow the other configuration's style.

sweisdb replied:
Good catch. Fixed.

sweisdb committed:
[SPARK-47318] Updating documentation

sweisdb committed:
[SPARK-47318] Fixing HTML table

mridulm commented on docs/security.md:
```diff
@@ -149,30 +149,44 @@ secret file agrees with the executors' secret file.
-is that data be encrypted in transit.
+The legacy method is an AES-based encryption mechanism relying on a shared secret. This requires
+RPC authentication to also be enabled. This method uses a bespoke protocol and it is recommended
+to use SSL instead.
```
Looking at the earlier text, we are dropping some verbiage from it - any particular reason for it ?
Specifically `"However, the AES based encryption is simpler to configure and may be preferred if the only requirement is that data be encrypted in transit."` for ex.

We have made our recommendation clear, and rest should be left to users to determine - and make the right tradeoff.

sweisdb replied:
I will take this out. I think it is rationalizing why there is an ad hoc protocol here instead in the first place. I agree that we should make it clear that TLS is strongly preferred and make it easy to migrate.

sweisdb replied:
I put the original text back in.

mridulm reviewed (commented):
Just one comment, rest looks good to me.

+CC @JoshRosen as well to have an extra pair of eyes to review.

mridulm commented:
Can you please fix the linter related build errors @sweisdb ? Thx

mridulm commented:
Btw, we would not be able to directly merge this PR into older versions - given the reference to SSL in the docs.
So unfortunately, we will need a follow up PR for 3.5.

sweisdb committed:
[SPARK-47318] Restoring original Security docs text

mridulm reviewed (commented):
Looks good to me, thanks for working on this @sweisdb !
+CC @dongjoon-hyun

mridulm reviewed (changes requested):
The java linter failures are still there, can you please fix those ?
See: https://github.com/sweisdb/spark/actions/runs/8603440358/job/23575468152

dongjoon-hyun reviewed (commented):
Ya, please use `dev/lint-java` script and fix the following lines, @sweisdb .
```
$ dev/lint-java
Using `mvn` from path: /opt/homebrew/bin/mvn
Using SPARK_LOCAL_IP=localhost
Checkstyle checks failed at following occurrences:
[ERROR] src/main/java/org/apache/spark/network/crypto/AuthEngine.java:[68] (sizes) LineLength: Line is longer than 100 characters (found 113).
[ERROR] src/main/java/org/apache/spark/network/util/TransportConf.java:[218] (sizes) LineLength: Line is longer than 100 characters (found 116).
[ERROR] src/main/java/org/apache/spark/network/util/TransportConf.java:[220] (sizes) LineLength: Line is longer than 100 characters (found 101).
[ERROR] src/test/java/org/apache/spark/network/crypto/AuthEngineSuite.java:[54] (sizes) LineLength: Line is longer than 100 characters (found 110).
[ERROR] src/test/java/org/apache/spark/network/crypto/AuthEngineSuite.java:[55] (sizes) LineLength: Line is longer than 100 characters (found 116).
```

sweisdb committed:
[SPARK-47318] Fixing lint issues

sweisdb commented:
> Ya, please use `dev/lint-java` script and fix the following lines, @sweisdb .
> 
> ```
> $ dev/lint-java
> Using `mvn` from path: /opt/homebrew/bin/mvn
> Using SPARK_LOCAL_IP=localhost
> Checkstyle checks failed at following occurrences:
> [ERROR] src/main/java/org/apache/spark/network/crypto/AuthEngine.java:[68] (sizes) LineLength: Line is longer than 100 characters (found 113).
> [ERROR] src/main/java/org/apache/spark/network/util/TransportConf.java:[218] (sizes) LineLength: Line is longer than 100 characters (found 116).
> [ERROR] src/main/java/org/apache/spark/network/util/TransportConf.java:[220] (sizes) LineLength: Line is longer than 100 characters (found 101).
> [ERROR] src/test/java/org/apache/spark/network/crypto/AuthEngineSuite.java:[54] (sizes) LineLength: Line is longer than 100 characters (found 110).
> [ERROR] src/test/java/org/apache/spark/network/crypto/AuthEngineSuite.java:[55] (sizes) LineLength: Line is longer than 100 characters (found 116).
> ```

Addressed the lint issues.

dongjoon-hyun reviewed (approved):
+1, LGTM (Pending CIs). Thank you!

sweisdb commented:
Looks like the CI build is passing: https://github.com/sweisdb/spark/actions/runs/8623329437

dongjoon-hyun commented:
Ya, it looks like that, @sweisdb .

dongjoon-hyun commented:
Hi, @mridulm . Could you switch your request change?

dongjoon-hyun commented:
BTW, @sweisdb , could you make two backporting PR to `branch-3.5` and `branch-3.4`?

I'm the release manager for Apache Spark 3.4.3 and targeting RC1 next Monday. I want to deliver this patch via Apache Spark 3.4.3.

sweisdb commented:
@dongjoon-hyun:
v3.5.0 backport here: https://github.com/apache/spark/pull/46014
v3.4.0 backport here: https://github.com/apache/spark/pull/46015

dongjoon-hyun commented:
Thank you so much!

mridulm commented:
Merged to master.
Thanks for fixing this @sweisdb !
Thanks for reviewing @dongjoon-hyun :-)
